### PR TITLE
[data/preprocessors] make one hot encoder return encoded vector

### LINKED ISSUE #48132: [Data/preprocessors] OneHotEncoder should return a single column with the one hot encoded value
### Description

Right now `OneHotEncoder` returns multiple columns, one per each unique value of the original column. This makes the result a lot more complicated to use, given that you will need to know all the unique values of the original column to select the right one hot encoded columns

### Use case

I would expect something like this:
```
ds = ray.data.from_pandas({"A": ["warm", "cold", "hot", "cold"]})
encoder = OneHotEncoder(columns=["A"])
encoder.fit_transform(ds)
#     A_cold  A_hot  A_warm  
# 0     0            0.           1
# 1.     1            0.           0   
# 2     0            1           0 
# 3     1            0.           0 

# but it would be great if it could be:
#        A  
# 0     [0, 0, 1]
# 1.     [1, 0, 1]   
# 2      [0, 1, 0]
# 3      [1, 0, 0]

### PR DESCRIPTION
**Pull request by martinbomio**
## Why are these changes needed?

https://github.com/ray-project/ray/issues/48132

## Related issue number

Closes https://github.com/ray-project/ray/issues/48132

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by martinbomio** [comment]
cc @richardliaw

File: python/ray/data/preprocessors/encoder.py [hunk 1]
```diff
@@ -230,8 +228,15 @@ def _transform_pandas(self, df: pd.DataFrame):
...
-        df = df.drop(columns=list(columns_to_drop))
+            # Concatenate the value columns
+            value_columns = [
+                f"{column}_{column_value}" for column_value in column_values
+            ]
+            concatenated = df[value_columns].to_numpy()
+            df = df.drop(columns=value_columns)
+            # Use a Pandas Series for column assignment to get more consistent
+            # behavior across Pandas versions.
+            df.loc[:, column] = pd.Series(list(concatenated))
```
[the remark this answers is not in the document]
Comment (martinbomio) [reply python/ray/data/preprocessors/encoder.py:239]: I think it should be fine to make it a numpy array but that would be inconsistent with other preprocessors. For example, the `MultiHotEncoder` returns a list, and the `Concatenator` also [returns](https://github.com/ray-project/ray/blob/master/python/ray/data/preprocessors/concatenator.py#L108) a list. I actually copy the code from the concatenator

[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/data/preprocessors/encoder.py:239]: ack let's move fwd for now.

**Review by richardliaw — APPROVED** [review]

**Review by omatthew98 — APPROVED** [review]
Lgtm

**Comment by richardliaw** [comment]
gracias

**Review by raulchen — APPROVED** [review]

**Pull request merged**